Add support for Closed Captions as per VAST 4.1 specification

leon486 opened the pull request:
Add support for <ClosedCaptionFiles> and <ClosedCaptionFile> tags used for Closed Captions as specified in VAST 4.1

leon486 committed:
Add support for Closed Captions as per VAST 4.1 specification

sokil commented on src/Creative/InLine/Linear/ClosedCaption.php:
```diff
@@ -0,0 +1,39 @@
+        $this->domElement = $domElement;
+    }
+    
+    public function setType($mime)
```
nee phpdocs with types

leon486 replied:
Added

sokil commented on src/Creative/InLine/Linear.php:
```diff
@@ -78,6 +80,42 @@ public function createMediaFile()
+    /**
+     * @return ClosedCaption
+     */
+    public function createClosedCaption()
```
need to add tests and  readme

leon486 replied:
Added

leon486 committed:
- Add tests for Closed Captions
- Add phpdocs for ClosedCaption class
- Add closed caption support section to readme
- Fix readme typos

leon486 committed:
Increase test coverage

leon486 committed:
Code readability improvements

leon486 committed:
Fixed test filename

leon486 commented:
Made the requested changes and all test passed.

sokil commented on src/Creative/InLine/Linear.php:
```diff
@@ -78,6 +80,42 @@ public function createMediaFile()
+     */
+    public function createClosedCaption()
+    {
+        if (empty($this->mediaFilesDomElement)) {
```
This code is copy paste from mediaFile builder and must be extracted

sokil commented on src/Creative/InLine/Linear.php:
```diff
@@ -78,6 +80,42 @@ public function createMediaFile()
+    /**
+     * @return ClosedCaption
+     */
+    public function createClosedCaption()
```
Specifaction calls this element ClosedCaptionFile

sokil commented on src/Creative/InLine/Linear.php:
```diff
@@ -78,6 +80,42 @@ public function createMediaFile()
+            }
+        }
+
+        if (empty($this->CCFilesDomElement)) {
```
Тo need to save on the symbols) CCFilesDomElement -> ClosedCaptionFilesDomElement

sokil commented on src/Creative/InLine/Linear/ClosedCaption.php:
```diff
@@ -0,0 +1,71 @@
+
+namespace Sokil\Vast\Creative\InLine\Linear;
+
+class ClosedCaption
```
ClosedCaptionFile

sokil commented on README.md:
````diff
@@ -118,14 +118,93 @@ This will generate:
+Closed Captions as per VAST 4.1 specification are supported as follows
+
+```php
+// create document
````
it's to much for that feaure ) you may just add your method to Quick start

leon486 committed:
- Move ClosedCaption class and file to ClosedCaptionFile
- Update Readme example
- Update ClosedCaptionFile usage
- Remove code clone in Linear.
- Update tests

leon486 committed:
Update example

leon486 committed:
- Use new createInLineAdLinearCreativeClosedCaptionFile method when creating closed caption objects

sokil commented on src/Creative/InLine/Linear.php:
```diff
@@ -12,6 +12,8 @@
 namespace Sokil\Vast\Creative\InLine;
 
 use Sokil\Vast\Creative\AbstractLinearCreative;
+
```
space between use

sokil commented on src/Creative/InLine/Linear.php:
```diff
@@ -53,22 +60,37 @@ public function setDuration($duration)
+     * @return VOID
+     */
+    private function createDomElement($domElement, $tagName, $parentTagName){
+      if (empty($this->{$domElement})) {
```
$this->{$domElement} - it's a magic, that hard to maintain and debug.  

You need a method that just created MediaFiles and returns so i think  `getMediaFilesElement` will be good.

```
$this->getMediaFilesElement()->createMediaFile();
$this->getMediaFilesElement()->createClosedCaptionFile();
```

leon486 replied:
The ClosedCaptionFile tag, needs to be within a ClosedCaptionFiles tag:
`<MediaFiles> <ClosedCaptionFiles> <ClosedCaptionFile>`

If I follow your idea, I would need to create a `getClosedCaptionFilesElement` method that would basically be a code clone of `getMediaFilesElement`.

like this:
`$this->getMediaFilesElement()->createMediaFile();`
`$this->getMediaFilesElement()->getClosedCaptionFilesElement()->createClosedCaptionFile();`

If I don't create the `getClosedCaptionFilesElement` method, then I would have to put the logic of getting the `<ClosedCaptionFiles>` DOM element within `createClosedCaptionFile`.

sokil replied:
When you copy-and-pase code that created MediaFiles element, this was code that creates same element, but writted in different places. 

Method `getClosedCaptionFilesElement` is not a clone of `getMediaFilesElement` because it needs first to create `MediaFiles` element. If code looks like similar, this does not mean thet code is similar. 

You do not need `getClosedCaptionFilesElement` because currently there is single place where this element may create - in `createClosedCaptionFile` method. 

```
public function createClosedCaptionFile() {
    if (empty($this->closedCaptionFilesDomElement) {
           $mediaFilesElement = $this->getMediaFilesElement()
            $this->closedCaptionFilesDomElement = $mediaFilesElement->getElementsByTagName('ClosedCaptionFiles')->item(0);	
            if (!$this->closedCaptionFilesDomElement) {	
                $this->closedCaptionFilesDomElement = $this->getDomElement()->ownerDocument->createElement('ClosedCaptionFiles');	
                $mediaFilesElement->appendChild($this->closedCaptionFilesDomElement);	
            }	
        }
}
```

Explicit is better than implicit.

leon486 replied:
You suggest a method that provides the MediaFilesDomElement to `getMediaFileElement` and `closedCaptionFileDomElement` that's the idea that I had in mind.

What I don't like too much is how similarly the `CaptionFilesDomElement` is provided to `closedCaptionFileDomElement`.

   ```
 private function getMediaFilesElement(){
      if (empty($this->mediaFilesDomElement)) {
          $this->mediaFilesDomElement = $this->getDomElement()->getElementsByTagName('MediaFiles')->item(0);
          if (!$this->mediaFilesDomElement) {
              $this->mediaFilesDomElement = $this->getDomElement()->ownerDocument->createElement('MediaFiles');
              $this->getDomElement()
                  ->getElementsByTagName('Linear')
                  ->item(0)
                  ->appendChild($this->mediaFilesDomElement);
          }
      }
      ...
    }
```

```
public function createClosedCaptionFile() 
    {
        //ensure closedCaptionFilesDomElement existence
        if (empty($this->closedCaptionFilesDomElement)) {
           $mediaFilesElement = $this->getMediaFilesElement();
            $this->closedCaptionFilesDomElement = $mediaFilesElement->getElementsByTagName('ClosedCaptionFiles')->item(0);	
            if (!$this->closedCaptionFilesDomElement) {	
                $this->closedCaptionFilesDomElement = $this->getDomElement()->ownerDocument->createElement('ClosedCaptionFiles');	
                $mediaFilesElement->appendChild($this->closedCaptionFilesDomElement);	
            }	
        }
        ...
    }
```
This is very similar code. But I don't mind.

leon486 committed:
- Remove empty line from Linear.php

leon486 committed:
- Pass the whole object instead of just the name to the DomElement creator function.

leon486 committed:
- Use getMediaFilesElement

sokil commented:
Merged. Pay attention on code formatting and PSR-1/PSR-2/PSR-12.
